sway-core/src/asm_generation/asm_builder.rs

```diff
@@ -607,16 +607,7 @@ impl<'ir> AsmBuilder<'ir> {
 
         // Index value is the array element index, not byte nor word offset.
         let index_reg = self.value_to_register(index_val);
-        let rel_offset_reg = match index_reg {
-            VirtualRegister::Virtual(_) => {
-                // We can reuse the register.
-                index_reg.clone()
-            }
-            VirtualRegister::Constant(_) => {
-                // We have a constant register, cannot reuse it.
-                self.reg_seqr.next()
-            }
-        };
+        let rel_offset_reg = self.reg_seqr.next();
```

Comment:
Oh, dear, yeah, it's almost surprising this problem didn't come up earlier.  I plan to refactor this so that `VirtualRegisters` can't be re-used and will behave a bit like SSA, so that we can do peephole opts without worrying about liveness.

Reply:
If we do this, then we may be able to simplify the register allocator a bit and make it faster. Liveness analysis currently **does not** assume SSA.

Reply:
Just to confirm, neither of you are suggesting any changes to this PR right?

This PR is ready, except that I haven't added the failing test to our testsuite. Once I'll do that I'll mark it as ready.

Reply:
Yeah we just like to talk clearly 🤣